streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala

```diff
@@ -0,0 +1,224 @@
+          failed = job.numFailedTasks, skipped = job.numSkippedTasks,
+          total = job.numTasks - job.numSkippedTasks)}
+        </td>
+        {failureReasonCell(lastFailureReason)}
+      </tr>
+    }
+
+    makeJobRow(jobs.head, true) ++ (jobs.tail.map(job => makeJobRow(job, false)).flatMap(x => x))
+  }
+
+  private def failureReasonCell(failureReason: String): Seq[Node] = {
+    val isMultiline = failureReason.indexOf('\n') >= 0
+    // Display the first line by default
+    val failureReasonSummary = StringEscapeUtils.escapeHtml4(
+      if (isMultiline) {
+        failureReason.substring(0, failureReason.indexOf('\n'))
+      } else {
+        failureReason
+      })
+    val details = if (isMultiline) {
+      // scalastyle:off
+      <span onclick="this.parentNode.querySelector('.stacktrace-details').classList.toggle('collapsed')"
+            class="expand-details">
+        +details
+      </span> ++
+        <div class="stacktrace-details collapsed">
+          <pre>{failureReason}</pre>
+        </div>
+      // scalastyle:on
+    } else {
+      ""
+    }
+    <td valign="middle" style="max-width: 300px">{failureReasonSummary}{details}</td>
+  }
+
+  private def jobsTable(jobInfos: Seq[(OutputOpId, JobId)]): Seq[Node] = {
+    def getJobData(jobId: JobId): Option[JobUIData] = {
+      sparkListener.activeJobs.get(jobId).orElse {
+        sparkListener.completedJobs.find(_.jobId == jobId).orElse {
+          sparkListener.failedJobs.find(_.jobId == jobId)
+        }
+      }
+    }
+
+    // Group jobInfos by OutputOpId firstly, then sort them.
+    // E.g., [(0, 1), (1, 3), (0, 2), (1, 4)] => [(0, [1, 2]), (1, [3, 4])]
+    val outputOpIdWithJobIds: Seq[(OutputOpId, Seq[JobId])] =
+      jobInfos.groupBy(_._1).toSeq.sortBy(_._1). // sorted by OutputOpId
+        map { case (outputOpId, jobs) =>
+        (outputOpId, jobs.map(_._2).sortBy(x => x).toSeq)} // sort JobIds for each OutputOpId
+    sparkListener.synchronized {
+      val outputOpIdWithJobs: Seq[(OutputOpId, Seq[JobUIData])] = outputOpIdWithJobIds.map {
+        case (outputOpId, jobIds) =>
+          // Filter out JobIds that don't exist in sparkListener
+          (outputOpId, jobIds.flatMap(getJobData))
+      }
+
+      <table id="batch-job-table" class="table table-bordered table-striped table-condensed">
+        <thead>
+          {columns}
+        </thead>
+        <tbody>
+          {outputOpIdWithJobs.map { case (outputOpId, jobs) => makeOutputOpIdRow(outputOpId, jobs)}}
+        </tbody>
+      </table>
+    }
+  }
+
+  def render(request: HttpServletRequest): Seq[Node] = {
+    val batchTime = Option(request.getParameter("id")).map(id => Time(id.toLong)).getOrElse {
+      throw new IllegalArgumentException(s"Missing id parameter")
+    }
+    val formattedBatchTime = UIUtils.formatDate(batchTime.milliseconds)
+    val (batchInfo, jobInfos) = streaminglistener.synchronized {
```

Comment:
Also, this is not job infos! This is only the outputId to jobId mapping. That was very confusing as in the Spark UI stuff there is StageInfo, JobUIData, StageUIData, and all sorts of classes which actually contain info.